### ISSUE TITLE
radssh exit error

### ISSUE THREAD

**Issue by tynin** [radssh/radssh#32]
I believe this started happening when I upgraded from Fedora 25 to 26.

If I radssh to some servers, and run nothing, I can exit (using Ctrl-D) cleanly, without error.

```
radssh 10.147.62.1-2
Connecting to 2 hosts...
..
RadSSH $ 
Shell exiting
```

However, if I run a command and then exit, I get an error… but it seems like the error doesn’t always occur…

```
radssh 10.147.62.1-2
Connecting to 2 hosts...
..
RadSSH $ ifconfig eth0 | grep "RX p"
[10.147.62.1]           RX packets:886472444 errors:0 dropped:0 overruns:0 frame:0
[10.147.62.2]           RX packets:864302772 errors:0 dropped:0 overruns:0 frame:0
Average completion time for 2 hosts: 0.102575s
RadSSH $ 
Shell exiting
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
IOError: [Errno 22] Invalid argument
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
IOError: [Errno 22] Invalid argument

```

Here are the versions of what radssh is currently using:
```
python -m radssh.__main__
RadSSH Main Module
Package RadSSH 1.1.1 from (/usr/lib/python2.7/site-packages/radssh/__init__.pyc)
  Using Paramiko  2.1.2 from /usr/lib/python2.7/site-packages/paramiko/__init__.pyc
  Using cryptography 1.7.2 from /usr/lib64/python2.7/site-packages/cryptography/__init__.pyc
  Using netaddr 0.7.19 from /usr/lib/python2.7/site-packages/netaddr/__init__.pyc

Python 2.7.13 (CPython)
Running on Linux [BCTDLXWAGNERRH]
  Fedora (26/Twenty Six)
Encoding for stdout: UTF-8

Checking runtime limits...
  System is able to open at least 10000 concurrent files
  File check completed in 0.014165 seconds

  System is able to run 10000 concurrent threads
  Thread check completed in 1.850287 seconds

End of runtime check

```

**Comment by radssh** [comment]
At a (debug) glance, the primary usage of `atexit` in the RadSSH code is to call the readline module `write_history_file` to save the shell command line history. There are a handful of other registered `atexit` calls that do not seem familiar, so I'm guessing that there's something wonky with the runtime environment that is interfering with the attempt to update `~/.radssh_history` (or whatever setting value for `historyfile` you may have.

```atexit.register(readline.write_history_file, histfile)```

Not sure why the raised exception is `Errno 22 - Invalid Argument`, but that's likely bubbling up from the readline module code, not something directly controlled by RadSSH. If you rename/delete the `~/.radssh_history` file, that might clear up the error, but I'm really curious what about that file state could be tripping up the `readline.write_history_file` so badly that it triggers such an exception.

**Comment by VinodKandasamy** [comment]
Ran into the same issue today. Seems to happen when the history length limit in readline is reached. Deleting ~/.radssh_history would solve the issue. Also, 
`import readline`
`readline.set_history_length(n)`
where `n` is some number larger than the previous length (number of lines in the file) solves the issue.

Looks like an issue with readline.write_history_file

**Comment by radssh** [comment]
Thanks for the additional info, Vinod!

Digging a little deeper, it seems that the GNU Readline (7.0) in Fedora 26/27 bleeds the errno on truncate to the Python readline module. Turns out it has nothing to do with registering or invoking atexit call on termination. This sample produces the error on Fedora 27 (readline 7.0.7), but not CentOS 7 (readline 6.2.10):
```
import readline

readline.set_history_length(5)

for x in range(10):
	readline.add_history(str(x))

readline.write_history_file('/tmp/issue32.history')
```
Does not occur with Python3 under Fedora27, and I suspect https://github.com/python/cpython/commit/c345ce1a69b3c4a46d87ef56d859bc70abfc74b4 may be the "fix", but possibly not backported to Python2.

**Comment by radssh** [comment]
Confirmed - https://bugs.python.org/issue10350 with last comment (won't bother backporting to 2.x).

Not sure how much unraveling is needed to properly fix this, but I'm tempted to band-aid it for the time being, if only to silence the exception on exit - it seems as though the exception occurs after the file contents are properly written (and truncated), but I don't want to mask any true errors on rewriting the history file if/when they do occur.

**Comment by radssh** [comment]
Released in 1.1.2
